pkg/container-collection/options.go

```diff
@@ -252,23 +255,164 @@ func WithPubSub(funcs ...pubsub.FuncNotify) ContainerCollectionOption {
+		Version:  gv.Version,
+		Resource: resKind,
+	}
+
+	res, err := dynamicClient.
+		Resource(params).
+		Namespace(resNamespace).
+		Get(context.TODO(), resName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("cannot fetch %s/%s %s/%s: %w",
+			resKind, resGroupVersion, resNamespace, resName, err)
+	}
+
+	return res.GetOwnerReferences(), nil
+}
+
+func ownerReferenceEnrichment(
+	containerDefinition *pb.ContainerDefinition,
+	ownerReferences []metav1.OwnerReference,
+) error {
+	dynamicClient, err := getKubeClientDynamic()
+	if err != nil {
+		return fmt.Errorf("failed get dynamic Kubernetes client: %w", err)
+	}
+
+	resGroupVersion := "v1"
+	resKind := "pods"
+	resName := containerDefinition.Podname
+	resNamespace := containerDefinition.Namespace
+
+	// Iterate until we reach the highest level of reference
+	for {
+		if len(ownerReferences) == 0 {
+			ownerReferences, err = getOwnerReference(dynamicClient,
+				resNamespace, resKind, resGroupVersion, resName)
+			if err != nil {
+				return fmt.Errorf("failed to get %s/%s/%s/%s owner reference: %w",
+					resNamespace, resKind, resGroupVersion, resName, err)
+			}
+
+			// No owner reference found
+			if len(ownerReferences) == 0 {
+				return nil
+			}
+		}
+
+		ownerRef := getExpectedOwnerReference(ownerReferences)
+		if ownerRef == nil {
+			// Any expected owner reference found
+			return nil
+		}
+
+		// Update container's owner reference
+		containerDefinition.OwnerReference = &pb.OwnerReference{
+			Apiversion: ownerRef.APIVersion,
+			Kind:       ownerRef.Kind,
+			Name:       ownerRef.Name,
+			Uid:        string(ownerRef.UID),
+		}
+
+		// Update parameters for next iteration (Namespace does not change)
+		resGroupVersion = containerDefinition.OwnerReference.Apiversion
+		resKind = strings.ToLower(containerDefinition.OwnerReference.Kind) + "s"
+		resName = containerDefinition.OwnerReference.Name
+		ownerReferences = nil
+	}
+}
+
 // WithKubernetesEnrichment automatically adds pod metadata
 //
 // ContainerCollection.ContainerCollectionInitialize(WithKubernetesEnrichment())
 func WithKubernetesEnrichment(nodeName string) ContainerCollectionOption {
 	return func(cc *ContainerCollection) error {
 		config, err := rest.InClusterConfig()
 		if err != nil {
-			return fmt.Errorf("cannot start Kubernetes client: %s", err)
+			return fmt.Errorf("cannot start Kubernetes client: %w", err)
 		}
 		clientset, err := kubernetes.NewForConfig(config)
 		if err != nil {
-			return fmt.Errorf("cannot start Kubernetes client: %s", err)
+			return fmt.Errorf("cannot start Kubernetes client: %w", err)
 		}
+
 		// Future containers
 		cc.containerEnrichers = append(cc.containerEnrichers, func(containerDefinition *pb.ContainerDefinition) bool {
-			// No need for this enricher if the data is already there
+			// Enrich only with owner reference if the data is already there
 			if containerDefinition.Podname != "" {
+				ownerReferenceEnrichment(containerDefinition, nil)
```

Comment:
IMO, it would be too much to drop the container. Currently, we are only dropping containers if they are paused or if we failed to set a fanotify watcher to be notified about its deletion. I don't think an error while enriching the conatiner with the owner reference information has the same weight, don't you think?

Reply:
Just logging is fine.